controllers/oneagent/oneagent_controller.go

```diff
@@ -125,79 +128,89 @@ func (r *ReconcileOneAgent) Reconcile(ctx context.Context, rec *utils.Reconcilia
 			return false, err
 		}
 	} else if err != nil {
 		return false, err
 	} else if hasDaemonSetChanged(dsDesired, dsActual) {
-		logger.Info("Updating existing daemonset")
+		rec.Log.Info("Updating existing daemonset")
 		if err = r.client.Update(ctx, dsDesired); err != nil {
 			return false, err
 		}
 	}
 
-	if instance.Status.OneAgent.Version == "" {
-		if instance.Status.OneAgent.UseImmutableImage && instance.Spec.OneAgent.Image == "" {
-			if instance.Spec.OneAgent.Version == "" {
+	if rec.Instance.Status.OneAgent.Version == "" {
+		if rec.Instance.Status.OneAgent.UseImmutableImage && rec.Instance.Spec.OneAgent.Image == "" {
+			if rec.Instance.Spec.OneAgent.Version == "" {
 				latest, err := dtc.GetLatestAgentVersion(dtclient.OsUnix, dtclient.InstallerTypeDefault)
 				if err != nil {
 					return false, fmt.Errorf("failed to get desired version: %w", err)
 				}
-				instance.Status.OneAgent.Version = latest
+				rec.Instance.Status.OneAgent.Version = latest
 			} else {
-				instance.Status.OneAgent.Version = instance.Spec.OneAgent.Version
+				rec.Instance.Status.OneAgent.Version = rec.Instance.Spec.OneAgent.Version
 			}
 		} else {
 			desired, err := dtc.GetLatestAgentVersion(dtclient.OsUnix, dtclient.InstallerTypeDefault)
 			if err != nil {
 				return false, fmt.Errorf("failed to get desired version: %w", err)
 			}
 
-			logger.Info("Updating version on OneAgent instance")
-			instance.Status.OneAgent.Version = desired
+			rec.Log.Info("Updating version on OneAgent instance")
+			rec.Instance.Status.OneAgent.Version = desired
 		}
 
-		instance.Status.SetPhase(dynatracev1alpha1.Deploying)
+		rec.Instance.Status.SetPhase(dynatracev1alpha1.Deploying)
 		updateCR = true
 	}
 
-	if instance.Status.Tokens != utils.GetTokensName(instance) {
-		instance.Status.Tokens = utils.GetTokensName(instance)
+	if rec.Instance.Status.Tokens != utils.GetTokensName(rec.Instance) {
+		rec.Instance.Status.Tokens = utils.GetTokensName(rec.Instance)
 		updateCR = true
 	}
 
 	return updateCR, nil
 }
 
-func (r *ReconcileOneAgent) getPods(ctx context.Context, instance *dynatracev1alpha1.DynaKube) ([]corev1.Pod, []client.ListOption, error) {
+func (r *ReconcileOneAgent) getDesiredDaemonSet(rec *utils.Reconciliation) (*appsv1.DaemonSet, error) {
+	kubeSysUID, err := kubesystem.GetUID(r.apiReader)
+	if err != nil {
+		return nil, err
+	}
+
+	dsDesired, err := newDaemonSetForCR(rec.Log, rec.Instance, r.fullStack, string(kubeSysUID), r.feature)
+	if err != nil {
+		return nil, err
+	}
+	return dsDesired, nil
+}
+
+func (r *ReconcileOneAgent) getPods(ctx context.Context, instance *dynatracev1alpha1.DynaKube, feature string) ([]corev1.Pod, []client.ListOption, error) {
 	podList := &corev1.PodList{}
 	listOps := []client.ListOption{
 		client.InNamespace((*instance).GetNamespace()),
-		client.MatchingLabels(buildLabels((*instance).GetName())),
+		client.MatchingLabels(buildLabels((*instance).GetName(), feature)),
```

Comment:
You don't need the dereference here,

```suggestion
		client.MatchingLabels(buildLabels(instance.Name, feature)),
```

Reply:
Done